src/Compilers/CSharp/Portable/Lowering/Instrumentation/ModuleCancellationInstrumenter.cs

```diff
@@ -0,0 +1,198 @@
+        => InstrumentStatement(base.InstrumentForEachStatementConditionalGotoStart(original, branchBack));
+
+    public override BoundStatement InstrumentGotoStatement(BoundGotoStatement original, BoundStatement rewritten)
+        => InstrumentStatement(base.InstrumentGotoStatement(original, rewritten));
+
+    public override void InterceptCallAndAdjustArguments(
+        ref MethodSymbol method,
+        ref BoundExpression? receiver,
+        ref ImmutableArray<BoundExpression> arguments,
+        ref ImmutableArray<RefKind> argumentRefKindsOpt,
+        bool invokedAsExtensionMethod,
+        SimpleNameSyntax? nameSyntax)
+    {
+        Previous.InterceptCallAndAdjustArguments(ref method, ref receiver, ref arguments, ref argumentRefKindsOpt, invokedAsExtensionMethod, nameSyntax);
+
+        if (arguments is [.., { Type: { } lastArgumentType } lastArgument] &&
+            (argumentRefKindsOpt.IsDefault || argumentRefKindsOpt is [.., RefKind.None]) &&
+            lastArgumentType.Equals(_throwMethod.ContainingType, TypeCompareKind.ConsiderEverything))
+        {
+            // The last argument is a CancellationToken. Replace it with the module-level token.
+            // Keep the previous expression so that side-effects are preserved.
+            arguments = [.. arguments[0..^1], _factory.Sequence([lastArgument], _factory.ModuleCancellationToken())];
+        }
+        else if (FindOverloadWithCancellationToken(method) is { } cancellableOverload)
+        {
+            // The method being invoked does not have a CancellationToken as the last parameter, but there is an overload that does.
+            // Invoke the other overload instead and pass in module-level token. 
+            method = cancellableOverload;
+            arguments = [.. arguments, _factory.ModuleCancellationToken()];
+            argumentRefKindsOpt = argumentRefKindsOpt.IsDefault ? default : [.. argumentRefKindsOpt, RefKind.None];
+        }
+    }
+
+    /// <summary>
+    /// Find an overload whose last parameter is <see cref="CancellationToken"/> and
+    /// the parameter types and ref kinds of all other parameters match those of <paramref name="method"/>.
+    /// </summary>
+    private MethodSymbol? FindOverloadWithCancellationToken(MethodSymbol method)
+    {
+        // It's unlikely that real-world APIs have overloads that differ in dynamic,
+        // but if they do avoid selecting them since their intended use might differ.
+        //
+        // Similarly, we also only consider overloads with the same visibility.
+        // We could potentially allow the visibility to differ,
+        // but we'd need to check if the overload is accessible at the call site.
+
+        const TypeCompareKind TypeComparisonKind = TypeCompareKind.CLRSignatureCompareOptions & ~TypeCompareKind.IgnoreDynamic;
+
+        foreach (var member in method.ContainingType.GetMembers(method.Name))
```

Comment:
>method.ContainingType [](http://example.com/codeflow?start=31&length=21)

I think we should enumerate members of containing type definition. Once we find the target method definition, we should "project" it back to the containing type, ```AsMember```, then construct if necessary.

Reply:
That is an option. The current approach handles the following case as well, where the specialized overloads are cancellable while the generic isn't:

```C#
void G<T>(T a);           
void G<T>(int a, CancellationToken token);
void G<T>(string a, CancellationToken token);
```

This wouldn't work with the other approach. Probably not super important, not sure if there are any APIs like this.

Reply:
> This wouldn't work with the other approach. 

That was understood. I think that only definitions should be used to detect compatible overloads. Assuming that generic specialization was part of the intent is a stretch to far, in my opinion.

Reply:
Actually, never mind. The above would only make sense if we also supported different arities, like:

```
void G<T>(T a);           
void G(int a, CancellationToken token);
void G(string a, CancellationToken token);
```

I think don't need to do this at this point.

Reply:
This class D<T>
{
    public static void G(int a) {}
    public static void G(T a, CancellationToken token) {}
}

class C
{
    void F(CancellationToken token)
    {
        D<int>.G(1);
    }
If an API has specialized cancellable version the current impl inserts the token, but if we enumerate the definition it wouldn't:

```C#
class D<T>
{
    public static void G(T a) {}
    public static void G(int a, CancellationToken token) {}
}

class C
{
    void F(CancellationToken token)
    {
        D<int>.G(1);
    }
}
```

Reply:
> If an API has specialized cancellable version the current impl inserts the token, but if we enumerate the definition it wouldn't

That was also understood. And I think that this is the behavior we should have. Overloading should be detected at the definition level.

Reply:
Similarly with generic methods, but seems less useful.

```C#
void F(CancellationToken token)
{
    G<int>(0, 1);
}

void G<T>(T a, T b) => throw null;
void G<T>(int a, T b, CancellationToken token) => throw null;
void G<T>(bool a, T b, CancellationToken token) => throw null;
```

Reply:
OK, I'll change it as you suggest.